developer_guides/uuid/index.rst

```diff
@@ -0,0 +1,133 @@
+introduced, the painpoints with this method including how to keep ABI
+aligned and backward compatible for different version combination of
+topology/driver/firmware for each new added components, there could be
+component type collision if versions are not aligned.
+
+UUIDs (Universally Unique Identifiers) provide a more scalable and
+collision-free way of component identification. UUIDs are currently
+used as the standard interface by all users of the SOF firmware,
+including the tracing subsystem, the topology .m4 files, and the Linux
+topology driver.
+
+Allocating a new UUID for the new added component is recommended, as
+the component type might be replaced by UUID in the future. For the
+whole SOF stack, the usage of UUID should follow rules as below:
+
+UUID Allocation
+***************
+The UUID allocation of a specific component should use the version 4 in
+`UUID wikipedia <https://en.wikipedia.org/wiki/Universally_unique_identifier>`__,
+and the value should be delared in the firmware component driver with
+``DECLARE_SOF_RT_UUID``, for details, please refer to the API
+documentation :ref:`uuid-api`.
+
+UUID in Topology
+****************
+We should provide the same UUID in topology .m4 file for a new added
+widget (corresponding to the new added component), since we have
+implmented marco to help to handle the converting task, just use the
+exactly same macro ``DECLARE_SOF_RT_UUID`` as depited in the firmware
+source, e.g for src component the below should be addded to the topology
+.m4 tools/topology/m4/src.m4:
+
+.. code-block:: none
+
+   DECLARE_SOF_RT_UUID("host", host_uuid, 0x8b9d100c, 0x6d78, 0x418f,
+        0x90, 0xa3, 0xe0, 0xe8, 0x05, 0xd0, 0x85, 0x2b);
+
+Linux Topology Driver
+*********************
+The topology driver will parse the 16Bytes UUID token, append it to the
+extended data of the IPC struct, and sent it to the
+firmware in component creation stage, **for all components**.
+
+.. code-block:: none
+
+   /* Component extended tokens */
+   static const struct sof_topology_token comp_ext_tokens[] = {
+    {SOF_TKN_COMP_UUID,
+        SND_SOC_TPLG_TUPLE_TYPE_UUID, get_token_uuid,
+        offsetof(struct sof_ipc_comp_new_ext, uuid), 0},
+   };
+
+.. code-block:: none
+
+   static int sof_widget_ready(struct snd_soc_component *scomp, int index,
+                               struct snd_soc_dapm_widget *w,
+                               struct snd_soc_tplg_dapm_widget *tw)
+   {
+   ...
+           ret = sof_parse_tokens(scomp, &swidget->comp_ext, comp_ext_tokens,
+                                  ARRAY_SIZE(comp_ext_tokens), tw->priv.array,
+                                  le32_to_cpu(tw->priv.size));
+   ...
+   }
+
+UUID Arrays Stored Section
+**************************
+Only the UUID arrays for runtime created components are stored to the
+.rodata section as static data, it will occupy small memory, e.g.
+19 component types * 16 Bytes/component type = 304 Bytes.
```

Comment:
this is unclear, this implies there are built-time components so where are those UUIDs stored?

Reply:
The wording 'runtime' is confused here, let me correct it.

Reply:
well this begs a follow-up question. how do we determine which components are used in topology files and which ones are not?

Reply:
Basically, the macro 'DECLARE_SOF_RT_UUID' will declare the corresponding UUID static data to the .rodata section. So, if a component(or module, or a pure driver only) is declared with macro 'DECLARE_SOF_RT_UUID' in the firmware, then its UUID should be specified in the topology. Otherwise, when declared with the macro 'DECLARE_SOF_UUID' only, it will be used for the system internal only, the tracing system will use it, but since its UUID will not located in the .rodata section, this will save the memory footprint (we need only 4 Bytes offset for each of this kind of definition).

If you grep the 'DECLARE_SOF_RT_UUID' usage in the firmware source, it is about 39 for the real use at this moment, if UUID of a component is defined in the firmware with 'DECLARE_SOF_RT_UUID', it will occupy 16 Bytes 
of the memory, no matter if the component will be used in your topology files or not.